examples/state_machine/states.py

```diff
@@ -0,0 +1,361 @@
+
+        return PNTransition(
+            state=HandshakeStoppedState,
+            context=self.context,
+            effect=effects.CancelHandshakeEffect()
+        )
+
+    def subscription_changed(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        self._context.channels = event.channels
+        self._context.groups = event.groups
+        self._context.tt = 0
+
+        if event.channels[0] == 'fail':
+            next_state = HandshakeReconnectingState \
+                if self._context.handshake_count < HandshakingState.MAX_HANDSHAKE_RETRY_COUNT \
+                else HandshakeFailedState
+        else:
+            next_state = ReceivingState
+
+        return PNTransition(
+            state=next_state
+        )
+
+    def handshake_reconnect(self, event: events.HandshakeReconnectFailureEvent, context: PNContext) -> PNTransition:
+        self._context = context.handshake_count + 1
+
+        if event.channels[0] == 'fail':
+            next_state = HandshakeReconnectingState \
+                if self._context.handshake_count < HandshakingState.MAX_HANDSHAKE_RETRY_COUNT \
+                else HandshakeFailedState
+        else:
+            next_state = ReceivingState
+
+        return PNTransition(
+            state=next_state,
+            context=self._context
+        )
+
+    def give_up(self, event: events.HandshakeReconnectGiveupEvent, context: PNContext) -> PNTransition:
+        return PNTransition(
+            state=HandshakeFailedState,
+            context=self._context
+        )
+
+    def restored(self, event: events.SubscriptionRestoredEvent, context: PNContext) -> PNTransition:
+        return PNTransition(
+            state=ReceivingState,
+            context=self._context
+        )
+
+    def success(self, event: events.HandshakeReconnectSuccessEvent, context: PNContext) -> PNTransition:
+        return PNTransition(
+            state=ReceivingState,
+            context=self._context
+        )
+
+
+class HandshakeFailedState(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+            events.HandshakeReconnectRetryEvent.__name__: self.reconnect_retry,
+            events.SubscriptionChangedEvent.__name__: self.subscription_changed,
+            events.ReconnectEvent.__name__: self.reconnect,
+            events.SubscriptionRestoredEvent.__name__: self.subscription_restored,
+        }
+
+    def reconnect_retry(self, event: events.HandshakeReconnectRetryEvent, context: PNContext) -> PNTransition:
+        return PNTransition(state=HandshakeReconnectingState)
+
+    def subscription_changed(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        return PNTransition(state=HandshakingState)
+
+    def reconnect(self, event: events.ReconnectEvent, context: PNContext) -> PNTransition:
+        return PNTransition(state=HandshakingState)
```

Comment:
I would expect some difference between `reconnect` and `subscription_changed`. One of them should take the channels and channel_groups from the event and the other from current context (as the reconnect event itself do not have that information), yet the implementations do not differ in the slightest. Which one should change?